Revert "Windows: Compile with /utf-8 flag"

sukibaby opened the pull request:
This reverts commit 58a45d91f4f4a62e5b0e5562fb0962409e4d27e4.

I am building with MSVC and tracked down an overall performance decrease between release and beta branch to this. A debugger will show this leads to inefficiencies anywhere RString is used.

sukibaby committed:
Revert "Windows: Compile with /utf-8 flag"

This reverts commit 58a45d91f4f4a62e5b0e5562fb0962409e4d27e4.

florczakraf commented:
>  an overall performance decrease between release and beta branch to this

could you share some numbers / benchmarks / hardware so others could test and confirm the impact?

teejusb commented:
Yeah I agree with @florczakraf. It's verrrryy not descriptive what you mean my "overall performance decrease". How much is the difference, and what is it affecting?

sukibaby commented:
@florczakraf 
@teejusb 

I won't go into a huge amount of detail, since one can compile in MSVC with and without the utf-8 flag, and you don't even have to play a song, you can just go through menus and see it's noticeably slower when built with utf-8. 

One can also follow along in x64dbg with both builds, but I hope this gets the point across:

You can tell easily from the main screen of Simply Love. I think this is fairly telling, especially on a high refresh rate monitor, of course being recorded by OBS and reconverted to mp4 is not an ideal way to show this but I think you can still see what's going on.

When the game is built with the utf-8 flag, the floating hearts stutter randomly, and there is a longer lag between each menu choice. Keep an eye on the hearts while I'm scrolling thru the color choices, for example, and the lag spikes shown by Rendering Stats are more pronounced.

Without the utf8 flag the song wheel and settings menu are both much smoother and the options menu also comes up quicker, in other words it's just worse all around even outside of gameplay.

Built with utf-8:


https://github.com/itgmania/itgmania/assets/163092272/f4c701c1-a3a2-439c-bf19-27a1daef1059




Current beta, built without utf-8:




https://github.com/itgmania/itgmania/assets/163092272/8d7c59d1-b0f0-4a30-9157-89027f2de5e3

florczakraf commented:
I did a blind test of the recordings you provided and couldn't really tell the difference between these two. Next, I asked people on discord to tell how they feel about them (1. 2nd video [build without /utf-8 flag] is smoother, 2. 1st is smoother, 3. no difference, both are not smooth, 4. no difference, both are smooth) without giving any context to not bias them. Out of 7 responses, or 8 including myself, no one had a preference for a particular video, so either they don't show the issue or there's no issue at all.

![image](https://github.com/itgmania/itgmania/assets/9034451/fe403516-969d-4503-a48b-7cf165bc1fb8)

---------

Could you please describe how does the flag in question affect the generated code and share measurable benchmarks? My current understanding is that compiling the game without this flag [makes the output depend on an arbitrary environment setting of the person compiling it](https://learn.microsoft.com/en-us/cpp/build/reference/execution-charset-set-execution-character-set?view=msvc-170). While having `/utf-8` doesn't resolve all issues around unicode, it sounds reasonable to keep the flag at least for consistency across builds

sukibaby commented:
> I did a blind test of the recordings you provided and couldn't really tell the difference between these two. Next, I asked people on discord to tell how they feel about them (1. 2nd video [build without /utf-8 flag] is smoother, 2. 1st is smoother, 3. no difference, both are not smooth, 4. no difference, both are smooth) without giving any context to not bias them. Out of 7 responses, or 8 including myself, no one had a preference for a particular video, so either they don't show the issue or there's no issue at all.
> 
> ![image](https://private-user-images.githubusercontent.com/9034451/335860186-fe403516-969d-4503-a48b-7cf165bc1fb8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTczOTI5NTgsIm5iZiI6MTcxNzM5MjY1OCwicGF0aCI6Ii85MDM0NDUxLzMzNTg2MDE4Ni1mZTQwMzUxNi05NjlkLTQ1MDMtYTQ4Yi03Y2YxNjViYzFmYjgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MDNUMDUzMDU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTk2YzVhOWIwZTJhYzhkOTYxMTYxMTYwMTFkYTNmOTg5Y2U3ZmFiN2ZlNzJmOTBiNGM1NTg5MWJkMmQ1NDkxNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.40DdV6QxLChSyV5BRuT_FvSc-82axar8ZUM4WVl2zwM)
> 
> Could you please describe how does the flag in question affect the generated code and share measurable benchmarks? My current understanding is that compiling the game without this flag [makes the output depend on an arbitrary environment setting of the person compiling it](https://learn.microsoft.com/en-us/cpp/build/reference/execution-charset-set-execution-character-set?view=msvc-170). While having `/utf-8` doesn't resolve all issues around unicode, it sounds reasonable to keep the flag at least for consistency across builds

On each of the screen recordings provided, I alt-tab in from OBS, let the lag value settle, and then go on from there. With utf-8 flag, there is a lag spike of 35ms going between menus which corresponds with a visible delay... this is easily noticeable by anyone. (There is a similar lag spike reported on the no-utf8 build but the display remains smooth). This is also recorded with OBS and then converted with HandBrake so it is far from an ideal way to show this issue. 

That is why I said anyone can simply try it with and without the flag, and find that it is simply slower all around. anyone can notice this if they try on both.

When I introduced this change to the builds provided to testers, the builds without /utf-8 were preferred to the builds that contained /utf-8 which I consider sufficient evidence.

If that doesn't satisfy you, give me a specific kind of measurement you want to see and I will try to accommodate that.  Do you want me to provide builds of the current beta with and without the MSVC utf-8 flag? 

I verified removing the flag doesn't break any actual utf-8 support in the game, and the flag is only being used when compiled with MSVC - there are already so many Windows ifdef's in the code and other Windows specifics, so I'm not sure that I agree on your point regarding "consistency"...?

florczakraf commented:
> When I introduced this change to the builds provided to testers, the builds without /utf-8 were preferred to the builds that contained /utf-8 which I consider sufficient evidence.

Was it an isolated change / were the builds including other changes as well?


> If that doesn't satisfy you, give me a specific kind of measurement you want to see and I will try to accommodate that. 

I think we'd need something that's easy to compare without any arbitrary bias. I don't know what's best here, a couple of things come up my mind when I see issues like this one: number of syscalls, CPU load, frame time stats, memory usage, FPS stats over time. Anything that would show the improvement objectively, across multiple test environments -- of course I don't expect you to provide results for more than yours, but rather a standardized way for people to reproduce and compare measurable results.


> Do you want me to provide builds of the current beta with and without the MSVC utf-8 flag?

I don't have access to a windows PC at the moment, I'll be back home next week -- I can build them myself.


> I verified removing the flag doesn't break any actual utf-8 support in the game, and the flag is only being used when compiled with MSVC - there are already so many Windows ifdef's in the code and other Windows specifics, so I'm not sure that I agree on your point regarding "consistency"...?

When talking about consistency I meant the build environment, not the ifdefs -- we all know that windows is a special snowflake that has to be treated with the utmost care. [Docs](https://learn.microsoft.com/en-us/cpp/build/reference/utf-8-set-source-and-executable-character-sets-to-utf-8?view=msvc-170) say (emphasis is mine): 

> By default, Visual Studio detects a byte-order mark to determine if the source file is in an encoded Unicode format, for example, UTF-16 or UTF-8. **If no byte-order mark is found, it assumes that the source file is encoded in the current user code page, unless you've specified a code page by using /utf-8 or the /source-charset option.** Visual Studio allows you to save your C++ source code in any of several character encodings. For information about source and execution character sets, see [Character sets](https://learn.microsoft.com/en-us/cpp/cpp/character-sets?view=msvc-170) in the language documentation.

There are plenty of source files where this applies in itgmania codebase, because they are non-ascii with mixed character encodings and no unicode BOM. Besides `UTF-8` there are also some `ISO-8859` ones:
```
(jorah)[~/proj/itgmania/src]
rf-> find . -type f -print0 | xargs -0 file | grep text | grep -v ASCII
./FontCharmaps.cpp:                                      C source, Unicode text, UTF-8 text
./NoteDisplay.cpp:                                       C source, Unicode text, UTF-8 text
./archutils/Win32/WindowsResources.in.rc:                C source, ISO-8859 text, with CRLF line terminators
./smpackage/ZipArchive/ChangeLog.txt:                    HTML document, ISO-8859 text, with very long lines (341)
./smpackage/ZipArchive/ZipArchive.rc:                    C source, ISO-8859 text, with CRLF line terminators
./smpackage/ZipArchive/ZipFileHeader.cpp:                C source, ISO-8859 text
./StepMania.xpm:                                         X pixmap image text, 32 x 32 x 310, 2 chars/pixel
./Profile.cpp:                                           C source, Unicode text, UTF-8 text, with CRLF line terminators
./StdString.h:                                           C++ source, Unicode text, UTF-8 text
./FontCharAliases.cpp:                                   C source, Unicode text, UTF-8 text
./ThemeManager.cpp:                                      C++ source, Unicode text, UTF-8 text
./Texture Font Generator/Texture Font Generator.rc:      C source, ISO-8859 text, with CRLF line terminators
./arch/Lights/LightsDriver_Win32Serial.h:                C++ source, Unicode text, UTF-8 text
./arch/Lights/LightsDriver_SextetStream.md:              Unicode text, UTF-8 text
./arch/Lights/SextetUtils.h:                             C source, Unicode text, UTF-8 text
./arch/Lights/LightsDriver_SextetStream.cpp:             C++ source, Unicode text, UTF-8 text
./arch/Lights/LightsDriver_SextetStream.h:               C++ source, Unicode text, UTF-8 text
./arch/InputHandler/InputHandler_SextetStream.md:        Unicode text, UTF-8 text, with CRLF line terminators
./arch/InputHandler/InputHandler_SextetStream.h:         C++ source, Unicode text, UTF-8 text
./arch/InputHandler/InputHandler_SextetStream.cpp:       C++ source, Unicode text, UTF-8 text
```

Exmple source files (I didn't check all of them) that would be handled using the code page set in the build environment, because they don't have BOM:
```
rf-> hexdump -C ./StdString.h | head -n 1
00000000  2f 2f 20 3d 3d 3d 3d 3d  3d 3d 3d 3d 3d 3d 3d 3d  |// =============|

rf-> hexdump -C ./FontCharmaps.cpp | head -n 1
00000000  23 69 6e 63 6c 75 64 65  20 22 67 6c 6f 62 61 6c  |#include "global|
```

Other [docs page](https://learn.microsoft.com/en-us/windows/win32/Intl/code-page-identifiers) suggests enforcing unicode for best portability:

> ANSI code pages can be different on different computers, or can be changed for a single computer, leading to data corruption. For the most consistent results, applications should use Unicode, such as UTF-8 or UTF-16, instead of a specific code page.

teejusb commented:
> With utf-8 flag, there is a lag spike of 35ms going between menus which corresponds with a visible delay

In your video itself you see a lag spike of 37ms without the utf-8 flag. 

Imo the videos don't really show a discernible difference in performance between the two. Comparing the values in the rendering stats definitely seem within margin of error.

sukibaby commented:
@teejusb 

> > With utf-8 flag, there is a lag spike of 35ms going between menus which corresponds with a visible delay
> 
> In your video itself you see a lag spike of 37ms without the utf-8 flag.
> 
> Imo the videos don't really show a discernible difference in performance between the two. Comparing the values in the rendering stats definitely seem within margin of error.

I really didn't intend for people to compare based off the videos alone.  I hoped to illustrate that if you try both EXE's, you can tell based off the main menu alone.

-----

@florczakraf 

> > When I introduced this change to the builds provided to testers, the builds without /utf-8 were preferred to the builds that contained /utf-8 which I consider sufficient evidence.
> 
> Was it an isolated change / were the builds including other changes as well?
>

It was an isolated change when I was testing myself. Since there was something between `release v0.8.0` and `beta` at the time I made the PR, and it was easy to tell if the problem was resolved by observing how choppy the main screen was, I decided I would isolate every commit made until the main screen stopped being choppy.

More specifically, when I was trying to track down the commit that impacted performance very negatively, I made a new branch based off of `beta` and individually reverted each commit.

 If a commit didn't change the choppiness of the main screen, I put it back, and went to the next one.
 
 So when I determined it was the fault of **"Windows: Compile with /utf-8 flag"**, I was using a branch of `beta` where that was the single only commit reverted, and the performance was exactly like it was on `release v0.8.0` again.
 
 When I put out test builds, it was not completely isolated- I believe the only other change I had implemented at the time was to reserve memory when `ActorMultiVertex` was resizing vectors, so it wasn't a perfectly isolated test, but that was a very very subtle change intended to fix a problem someone was having with performance when Step Statistics was enabled.


> > If that doesn't satisfy you, give me a specific kind of measurement you want to see and I will try to accommodate that.
> 
> I think we'd need something that's easy to compare without any arbitrary bias. I don't know what's best here, a couple of things come up my mind when I see issues like this one: number of syscalls, CPU load, frame time stats, memory usage, FPS stats over time. Anything that would show the improvement objectively, across multiple test environments -- of course I don't expect you to provide results for more than yours, but rather a standardized way for people to reproduce and compare measurable results.

It's a little hard because since SM5 / ITGm is mostly single-threaded, we can't use profilers - or I'd have provided a quick profiling of both programs. 

It would be possible for me to follow along both builds in x64dbg, but if it's necessary, I have a few ideas, since that debugger has some nice features to watch specific memory locations or watch for certain instructions.

> > Do you want me to provide builds of the current beta with and without the MSVC utf-8 flag?
> 
> I don't have access to a windows PC at the moment, I'll be back home next week -- I can build them myself.

👍 

> > I verified removing the flag doesn't break any actual utf-8 support in the game, and the flag is only being used when compiled with MSVC - there are already so many Windows ifdef's in the code and other Windows specifics, so I'm not sure that I agree on your point regarding "consistency"...?
> 
> When talking about consistency I meant the build environment, not the ifdefs -- we all know that windows is a special snowflake that has to be treated with the utmost care. [Docs](https://learn.microsoft.com/en-us/cpp/build/reference/utf-8-set-source-and-executable-character-sets-to-utf-8?view=msvc-170) say (emphasis is mine):
> 
> > By default, Visual Studio detects a byte-order mark to determine if the source file is in an encoded Unicode format, for example, UTF-16 or UTF-8. **If no byte-order mark is found, it assumes that the source file is encoded in the current user code page, unless you've specified a code page by using /utf-8 or the /source-charset option.** Visual Studio allows you to save your C++ source code in any of several character encodings. For information about source and execution character sets, see [Character sets](https://learn.microsoft.com/en-us/cpp/cpp/character-sets?view=msvc-170) in the language documentation.
> 
> There are plenty of source files where this applies in itgmania codebase, because they are non-ascii with mixed character encodings and no unicode BOM. Besides `UTF-8` there are also some `ISO-8859` ones:
> 
> ```
> (jorah)[~/proj/itgmania/src]
> rf-> find . -type f -print0 | xargs -0 file | grep text | grep -v ASCII
> ./FontCharmaps.cpp:                                      C source, Unicode text, UTF-8 text
> ./NoteDisplay.cpp:                                       C source, Unicode text, UTF-8 text
> ./archutils/Win32/WindowsResources.in.rc:                C source, ISO-8859 text, with CRLF line terminators
> ./smpackage/ZipArchive/ChangeLog.txt:                    HTML document, ISO-8859 text, with very long lines (341)
> ./smpackage/ZipArchive/ZipArchive.rc:                    C source, ISO-8859 text, with CRLF line terminators
> ./smpackage/ZipArchive/ZipFileHeader.cpp:                C source, ISO-8859 text
> ./StepMania.xpm:                                         X pixmap image text, 32 x 32 x 310, 2 chars/pixel
> ./Profile.cpp:                                           C source, Unicode text, UTF-8 text, with CRLF line terminators
> ./StdString.h:                                           C++ source, Unicode text, UTF-8 text
> ./FontCharAliases.cpp:                                   C source, Unicode text, UTF-8 text
> ./ThemeManager.cpp:                                      C++ source, Unicode text, UTF-8 text
> ./Texture Font Generator/Texture Font Generator.rc:      C source, ISO-8859 text, with CRLF line terminators
> ./arch/Lights/LightsDriver_Win32Serial.h:                C++ source, Unicode text, UTF-8 text
> ./arch/Lights/LightsDriver_SextetStream.md:              Unicode text, UTF-8 text
> ./arch/Lights/SextetUtils.h:                             C source, Unicode text, UTF-8 text
> ./arch/Lights/LightsDriver_SextetStream.cpp:             C++ source, Unicode text, UTF-8 text
> ./arch/Lights/LightsDriver_SextetStream.h:               C++ source, Unicode text, UTF-8 text
> ./arch/InputHandler/InputHandler_SextetStream.md:        Unicode text, UTF-8 text, with CRLF line terminators
> ./arch/InputHandler/InputHandler_SextetStream.h:         C++ source, Unicode text, UTF-8 text
> ./arch/InputHandler/InputHandler_SextetStream.cpp:       C++ source, Unicode text, UTF-8 text
> ```
> 
> Exmple source files (I didn't check all of them) that would be handled using the code page set in the build environment, because they don't have BOM:
> 
> ```
> rf-> hexdump -C ./StdString.h | head -n 1
> 00000000  2f 2f 20 3d 3d 3d 3d 3d  3d 3d 3d 3d 3d 3d 3d 3d  |// =============|
> 
> rf-> hexdump -C ./FontCharmaps.cpp | head -n 1
> 00000000  23 69 6e 63 6c 75 64 65  20 22 67 6c 6f 62 61 6c  |#include "global|
> ```
> 
> Other [docs page](https://learn.microsoft.com/en-us/windows/win32/Intl/code-page-identifiers) suggests enforcing unicode for best portability:
> 
> > ANSI code pages can be different on different computers, or can be changed for a single computer, leading to data corruption. For the most consistent results, applications should use Unicode, such as UTF-8 or UTF-16, instead of a specific code page.

This point further illustrates my concern that we don't know _what_ the compiler is doing differently to non UTF-8 files with the `/utf-8` flag, so I suggest reverting it if only because we don't know what effects it has on the codebase as a whole at this time (besides that it introduces small amounts of lag all around). 

Funny that you noted `StdString.h`, as that is one of the most awful pieces of code holding the SM5 code base back... I've spent so much time trying to make a replacement for that file that's C++14 or better but nothing works 😠 💢 ‼️ 

I do agree that everything should be Unicode, but I would like to make sure everything is working as expected too :)

teejusb commented:
I think I'm okay merging this PR, and then working to fix the files and maybe later reintroduce once we know things are good.
